src/platform/tls_schannel.c

```diff
@@ -1586,6 +1586,83 @@ CxPlatTlsUninitialize(
     }
 }
 
+_IRQL_requires_max_(PASSIVE_LEVEL)
+CXPLAT_TLS_RESULT_FLAGS
+CxPlatTlsIndicateCertificateReceived(
+    _In_ CXPLAT_TLS* TlsContext,
+    _In_ CXPLAT_TLS_PROCESS_STATE* State,
+    _In_ SecPkgContext_CertificateValidationResult* CertValidationResult
+    )
+{
+    SECURITY_STATUS SecStatus;
+    CXPLAT_TLS_RESULT_FLAGS Result = 0;
+    QUIC_CERTIFICATE* Certificate = NULL;
+    QUIC_CERTIFICATE_CHAIN* CertificateChain = NULL;
+
+#ifdef _KERNEL_MODE
+    SecPkgContext_Certificates PeerCert;
+    CxPlatZeroMemory(&PeerCert, sizeof(PeerCert));
+    SecStatus =
+        QueryContextAttributesW(
+            &TlsContext->SchannelContext,
+            SECPKG_ATTR_REMOTE_CERTIFICATES,
+            (PVOID)&PeerCert);
+#else
+    PCCERT_CONTEXT PeerCert = NULL;
+    SecStatus =
+        QueryContextAttributesW(
+            &TlsContext->SchannelContext,
+            SECPKG_ATTR_REMOTE_CERT_CONTEXT,
+            (PVOID)&PeerCert);
+#endif
+    if (SecStatus != SEC_E_OK) {
+        QuicTraceEvent(
+            TlsErrorStatus,
+            "[ tls][%p] ERROR, %u, %s.",
+            TlsContext->Connection,
+            SecStatus,
+            "Query peer cert");
+    } else {
+#ifdef _KERNEL_MODE
+        Certificate = (QUIC_CERTIFICATE*)&PeerCert;
+        CertificateChain = (QUIC_CERTIFICATE_CHAIN*)&PeerCert;
+#else
+        CXPLAT_DBG_ASSERT(PeerCert != NULL);
+        Certificate = (QUIC_CERTIFICATE*)PeerCert;
+        CertificateChain = (QUIC_CERTIFICATE_CHAIN*)(PeerCert->hCertStore);
```

Comment:
```suggestion
        CertificateChain = (QUIC_CERTIFICATE_CHAIN*)(PeerCert ? PeerCert->hCertStore : NULL);
```

Reply:
In this case, this is ok. This only occurs when the query call returns success, and the contract there says it returns a valid cert if successful.

Reply:
I could theoretically remove the dbg assert too, but I'd really like to know if schannel ever breaks that contract.

Reply:
Oh, now I see what that you did there. Makes sense.  These `#ifdef`s mess up readability, IMO (but there's no alternative).